pkg/controller/controller_utils.go

```diff
@@ -0,0 +1,291 @@
+	}
+	// TODO: Version this serialization per #7322
+	createdByRefJson, err := json.Marshal(createdByRef)
+	if err != nil {
+		return fmt.Errorf("unable to serialize controller reference: %v", err)
+	}
+
+	desiredAnnotations[CreatedByAnnotation] = string(createdByRefJson)
+
+	// use the dash (if the name isn't too long) to make the pod name a bit prettier
+	prefix := fmt.Sprintf("%s-", controller.Name)
+	if ok, _ := validation.ValidatePodName(prefix, true); !ok {
+		prefix = controller.Name
+	}
+
+	pod := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Labels:       desiredLabels,
+			Annotations:  desiredAnnotations,
+			GenerateName: prefix,
+		},
+	}
+	if err := api.Scheme.Convert(&controller.Spec.Template.Spec, &pod.Spec); err != nil {
+		return fmt.Errorf("unable to convert pod template: %v", err)
+	}
+	if labels.Set(pod.Labels).AsSelector().Empty() {
+		return fmt.Errorf("unable to create pod replica, no labels")
+	}
+	if newPod, err := r.kubeClient.Pods(namespace).Create(pod); err != nil {
+		r.recorder.Eventf(controller, "failedCreate", "Error creating: %v", err)
+		return fmt.Errorf("unable to create pod replica: %v", err)
+	} else {
+		glog.V(4).Infof("Controller %v created pod %v", controller.Name, newPod.Name)
+		r.recorder.Eventf(controller, "successfulCreate", "Created pod: %v", newPod.Name)
+	}
+	return nil
+}
+
+func (r RealPodControl) deletePod(namespace, podID string) error {
+	return r.kubeClient.Pods(namespace).Delete(podID)
+}
+
+// activePods type allows custom sorting of pods so an rc can pick the best ones to delete.
+type activePods []*api.Pod
+
+func (s activePods) Len() int      { return len(s) }
+func (s activePods) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s activePods) Less(i, j int) bool {
+	// Unassigned < assigned
+	if s[i].Spec.Host == "" && s[j].Spec.Host != "" {
+		return true
+	}
+	// PodPending < PodUnknown < PodRunning
+	m := map[api.PodPhase]int{api.PodPending: 0, api.PodUnknown: 1, api.PodRunning: 2}
+	if m[s[i].Status.Phase] != m[s[j].Status.Phase] {
+		return m[s[i].Status.Phase] < m[s[j].Status.Phase]
+	}
+	// Not ready < ready
+	if !api.IsPodReady(s[i]) && api.IsPodReady(s[j]) {
+		return true
+	}
+	return false
+}
+
+// filterActivePods returns pods that have not terminated.
+func filterActivePods(pods []api.Pod) []*api.Pod {
+	var result []*api.Pod
+	for i := range pods {
+		if api.PodSucceeded != pods[i].Status.Phase &&
+			api.PodFailed != pods[i].Status.Phase {
+			result = append(result, &pods[i])
+		}
+	}
+	return result
+}
+
+// updateReplicaCount attempts to update the Status.Replicas of the given controller, with retries.
+// Note that the controller pointer might contain a more recent version of the same controller passed into the function.
+func updateReplicaCount(rcClient client.ReplicationControllerInterface, controller *api.ReplicationController, numReplicas int) error {
+	return wait.Poll(10*time.Millisecond, 100*time.Millisecond, func() (bool, error) {
```

Comment:
How about try twice, and if it doesn't work the second time then re-queue? I'm just thinking that stuff will change--the RC might already be dirty again--so we shouldn't try to update it for too long.

This can come in a follow-up PR I think...